requests/models.py

```diff
@@ -470,18 +469,12 @@ def prepare_body(self, data, files, json=None):
         self.body = body
 
     def prepare_content_length(self, body):
-        if hasattr(body, 'seek') and hasattr(body, 'tell'):
-            curr_pos = body.tell()
-            body.seek(0, 2)
-            end_pos = body.tell()
-            self.headers['Content-Length'] = builtin_str(max(0, end_pos - curr_pos))
-            body.seek(curr_pos, 0)
-        elif body is not None:
-            l = super_len(body)
-            if l:
-                self.headers['Content-Length'] = builtin_str(l)
-        elif (self.method not in ('GET', 'HEAD')) and (self.headers.get('Content-Length') is None):
-            self.headers['Content-Length'] = '0'
+        if self.headers.get('Content-Length') is None:
+            if body is not None:
+                length = super_len(body)
+                self.headers['Content-Length'] = builtin_str(length)
+            elif self.method not in ('GET', 'HEAD'):
+                self.headers['Content-Length'] = '0'
```

Comment:
It's been a while since I was in this, so I've lost track of some of my thought process but it's slowly coming back. While I agree that it's pretty easy to accidentally shoot yourself in the foot with this approach, we specifically approved an odd subset of this functionality in #2329.

```python
>>> import requests
>>> r = requests.Request('POST', 'http://httpbin.org/post', headers={'Content-Length': '5'})
>>> r.headers
{'Content-Length': '5'}
>>> pr = r.prepare()
>>> pr.headers
{'Content-Length': '5'}
```

Given the above reasoning, why is this acceptable here but not when a body is attached? I don't mean to waste your time explaining this to me, I just think there's enough ambiguity in the implementation here that I want to make sure reasoning is fully understood before we lay this to rest.